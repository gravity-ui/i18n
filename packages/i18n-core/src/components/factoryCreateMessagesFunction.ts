import {memoize} from '@formatjs/fast-memoize';
import {type MessageDescriptor, MissingTranslationError} from '@formatjs/intl';

import type {MessageValue, Messages, MultiLocaleMessage} from '@gravity-ui/i18n-types';
import {ResolvedIntlConfigWithLocale} from '../types';
import {getFallbackLocalesList} from '../utils/fallbacks';
import {isLegacyPlural, convertLegacyPluralToICU} from '@gravity-ui/i18n-types';

const isMessageDefined = (message: MessageValue | undefined): message is MessageValue => {
    return message !== null && typeof message !== 'undefined';
};

export function factoryCreateMessagesFunction<TBase, AvailableLocale extends string>(
    config: Pick<
        ResolvedIntlConfigWithLocale<TBase, AvailableLocale>,
        | 'getLocale'
        | 'fallbackLocales'
        | 'disableUseLocaleLangAsFallback'
        | 'allowedLocales'
        | 'defaultFallback'
    >,
) {
    return function createMessages<K extends string>(
        msgs: Messages<AvailableLocale, K>,
    ): Record<K, MessageDescriptor> {
        // TODO Optimize?
        const getMessageDescriptor = memoize(
            ({
                key,
                message,
                currentLocale,
            }: {
                key: string | symbol;
                message: Messages<AvailableLocale, K>[K];
                currentLocale: string;
            }) => {
                const messages = message as MultiLocaleMessage<AvailableLocale>;
                let defaultMessage: MessageValue | undefined =
                    messages?.[currentLocale as AvailableLocale];

                if (!isMessageDefined(defaultMessage)) {
                    // TODO make cache
                    const fallbackLocales = getFallbackLocalesList(currentLocale, config);
                    const fallbackLocale = fallbackLocales.find((locale) =>
                        isMessageDefined(messages?.[locale as AvailableLocale]),
                    );
                    if (fallbackLocale) {
                        defaultMessage = messages?.[fallbackLocale as AvailableLocale];
                    }
                }

                if (!isMessageDefined(defaultMessage)) {
                    if (config.defaultFallback === 'key') {
                        defaultMessage = String(key);
                    } else if (config.defaultFallback === 'empty-string') {
                        defaultMessage = '';
                    } else {
                        throw new MissingTranslationError(
                            {
                                id: String(key),
                            },
                            config.getLocale(),
                        );
                    }
                }

                if (!defaultMessage) {
                    defaultMessage = '';
                }

                if (isLegacyPlural(defaultMessage)) {
                    defaultMessage = convertLegacyPluralToICU(defaultMessage);
                }

                const id = message?.meta?.id || defaultMessage || String(key);

                return {
                    // TODO Think about what to put to error. Use defaultMessage as ID?
                    id,
                    defaultMessage,
                } as MessageDescriptor;
            },
        );

        return new Proxy({} as Record<K, MessageDescriptor>, {
            get(_target, key) {
                const baseMsg = msgs[key as K];

                if (typeof baseMsg === 'undefined') {
                    throw new Error(`Not found message for key ${String(key)}`);
                }

                return getMessageDescriptor({
                    key,
                    message: baseMsg,
                    currentLocale: config.getLocale(),
                });
            },
        });
    };
}
