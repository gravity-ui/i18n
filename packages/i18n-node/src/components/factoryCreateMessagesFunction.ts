/// <reference types="intl-messageformat" />
import type {IntlFormattersFull} from '@gravity-ui/i18n-core';
import type {Messages} from '@gravity-ui/i18n-types';
import {
    factoryCreateMessagesFunction as coreFactoryCreateMessagesFunction,
    factoryCreateMessageFunctions,
} from '@gravity-ui/i18n-core';

import type {ResolvedIntlConfigWithLocale} from '../types';

export function factoryCreateMessagesFunction<AvailableLocale extends string>(
    config: ResolvedIntlConfigWithLocale<AvailableLocale>,
    intlFormatters: Pick<IntlFormattersFull<string>, 'formatMessage'>,
) {
    return function createMessages<K extends string>(msgs: Messages<AvailableLocale, K>) {
        const coreCreateMessages = coreFactoryCreateMessagesFunction(config);
        const messages = coreCreateMessages(msgs);
        const createMessageFunctions = factoryCreateMessageFunctions(intlFormatters);

        return createMessageFunctions(messages);
    };
}
