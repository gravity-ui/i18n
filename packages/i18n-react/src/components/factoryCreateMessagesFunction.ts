import type {Messages} from '@gravity-ui/i18n-types';
import {factoryCreateMessagesFunction as coreFactoryCreateMessagesFunction} from '@gravity-ui/i18n-core';
import type {IntlFormatters} from 'react-intl';

import type {ResolvedIntlConfig} from '../types';

import {factoryCreateMessageFunctions} from './factoryCreateMessageFunctions';

export function factoryCreateMessagesFunction<AvailableLocale extends string>(
    config: ResolvedIntlConfig<AvailableLocale>,
    intlFormatters: Pick<IntlFormatters<React.ReactNode>, 'formatMessage'>,
) {
    return function createMessages<K extends string>(msgs: Messages<AvailableLocale, K>) {
        const coreCreateMessages = coreFactoryCreateMessagesFunction(config);
        const messages = coreCreateMessages(msgs);
        const createMessageFunctions = factoryCreateMessageFunctions(config, intlFormatters);

        return {
            messages,
            ...createMessageFunctions(messages),
        };
    };
}
