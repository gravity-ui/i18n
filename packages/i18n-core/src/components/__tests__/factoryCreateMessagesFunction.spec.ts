import {factoryCreateMessagesFunction} from '../factoryCreateMessagesFunction';

describe('factoryCreateMessagesFunction', () => {
    it('1', () => {
        const createMessagesF = factoryCreateMessagesFunction({
            allowedLocales: ['en', 'ru'],
            getLocale: () => 'en',
            fallbackLocales: {},
            defaultFallback: 'en',
            disableUseLocaleLangAsFallback: true,
        });

        const messages = createMessagesF({
            create: {
                en: 'Create',
                ru: 'Создать',
            },
        });

        expect(messages['create']).toEqual({
            defaultMessage: 'Create',
            id: 'Create',
        });
    });

    it('2', () => {
        const createMessagesF = factoryCreateMessagesFunction({
            allowedLocales: ['en'],
            getLocale: () => 'en',
            fallbackLocales: {},
            defaultFallback: 'en',
            disableUseLocaleLangAsFallback: true,
        });

        const messages = createMessagesF({
            create: {
                en: 'Create',
                meta: {
                    id: 'create-id',
                },
            },
        });

        expect(messages['create']).toEqual({
            defaultMessage: 'Create',
            id: 'create-id',
        });
    });

    it('3', () => {
        const createMessagesF = factoryCreateMessagesFunction({
            allowedLocales: ['en'],
            getLocale: () => 'unknown',
            fallbackLocales: {},
            defaultFallback: 'empty-string',
            disableUseLocaleLangAsFallback: true,
        });

        const messages = createMessagesF({
            create: {
                en: 'Create',
            },
        });

        expect(messages['create']).toEqual({
            defaultMessage: '',
            id: 'create',
        });
    });

    it('4', () => {
        const createMessagesF = factoryCreateMessagesFunction({
            allowedLocales: ['en'],
            getLocale: () => 'unknown',
            fallbackLocales: {},
            defaultFallback: 'key',
            disableUseLocaleLangAsFallback: true,
        });

        const messages = createMessagesF({
            create: {
                en: 'Create',
            },
        });

        expect(messages['create']).toEqual({
            defaultMessage: 'create',
            id: 'create',
        });
    });

    it('5', () => {
        const createMessagesF = factoryCreateMessagesFunction({
            allowedLocales: ['en'],
            getLocale: () => 'unknown',
            fallbackLocales: {},
            defaultFallback: 'en',
            disableUseLocaleLangAsFallback: true,
        });

        const messages = createMessagesF({
            create: {
                en: null,
            },
        });

        try {
            // eslint-disable-next-line @typescript-eslint/no-unused-expressions
            messages['create'];
        } catch (e) {
            expect(e).toMatchSnapshot('5');
        }
    });

    it('6', () => {
        const createMessagesF = factoryCreateMessagesFunction({
            allowedLocales: ['en'],
            getLocale: () => 'unknown',
            fallbackLocales: {},
            defaultFallback: 'en',
            disableUseLocaleLangAsFallback: true,
        });

        const messages = createMessagesF({
            create: {
                en: null,
            },
        });

        try {
            // @ts-ignore
            // eslint-disable-next-line @typescript-eslint/no-unused-expressions
            messages['unknown'];
        } catch (e) {
            expect(String(e)).toMatchSnapshot('6');
        }
    });
});
