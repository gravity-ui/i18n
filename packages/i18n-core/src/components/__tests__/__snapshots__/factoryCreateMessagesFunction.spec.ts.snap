// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factoryCreateMessagesFunction 5: 5 1`] = `
[Error: [@formatjs/intl Error MISSING_TRANSLATION] Missing message: "create" for locale "unknown", using id as fallback.
]
`;

exports[`factoryCreateMessagesFunction 6: 6 1`] = `"Error: Not found message for key unknown"`;
